crates/networking/p2p/sync/state_sync.rs

```diff
@@ -241,8 +243,9 @@ impl StateSyncProgress {
     async fn update_key(progress: StateSyncProgress, segment_number: usize, current_key: H256) {
         progress.data.lock().await.current_keys[segment_number] = current_key
     }
-    async fn end_segment(progress: StateSyncProgress, segment_number: usize) {
-        progress.data.lock().await.ended[segment_number] = true
+    async fn end_segment(progress: StateSyncProgress, segment_number: usize, last_key: H256) {
+        progress.data.lock().await.ended[segment_number] = true;
+        progress.data.lock().await.current_keys[segment_number] = last_key;
```

Comment:
You probably want to do this holding the lock for both values together:
```suggestion
        let data = progress.data.lock().await;
        data.ended[segment_number] = true;
        data.current_keys[segment_number] = last_key;
```
That way you can keep both values synchronized at all times (otherwise, another task could read the segment ended but find an older key). It also saves a little bit of overhead, but that's probably negligible.

Reply:
Nice catch!